### PR TITLE
Add support for static networking in the autoyast templates.

### PR DESCRIPTION
**Pull request by hboetes**
I noticed the Suse based installs didn't support static networking, unlike debian and redhat based systems, and I needed it for my project.

Do note I basically applied the same diff twice to two different templates.

### CODE REVIEW

File: provisioning_templates/provision/autoyast_default.erb [hunk 1]
```diff
@@ -22,21 +22,54 @@ oses:
   <networking>
     <dns>
       <hostname><%= @host.name %></hostname>
+      <% if @host.subnet.respond_to?(:dhcp_boot_mode?) -%>
+      <dhcp_hostname config:type="boolean">false</dhcp_hostname>
+      <dhcp_resolv config:type="boolean">false</dhcp_resolv>
+      <domain><%= @host.domain %></domain>
+      <nameservers config:type="list">
+        <nameserver><%= @host.subnet.dns_primary %></nameserver>
+        <nameserver><%= @host.subnet.dns_secondary %></nameserver>
```
Comment (Klaas-) [inline provisioning_templates/provision/autoyast_default.erb]: I think we should check for presence of those nameservers, at least the secondary one - check the other templates: something like this:
<% if @host.subnet.dns_secondary.present? -%>
        <nameserver><%= @host.subnet.dns_secondary %></nameserver>
<% end -%>

File: provisioning_templates/provision/autoyast_default.erb [hunk 2]
```diff
@@ -22,21 +22,54 @@ oses:
...
+    <% end -%>
     </dns>
     <interfaces config:type="list">
       <interface>
-        <bootproto>dhcp</bootproto>
         <device>eth0</device>
+        <ethtool_options>autoneg on</ethtool_options>
+        <% if @host.subnet.respond_to?(:dhcp_boot_mode?) -%>
+          <bootproto>static</bootproto>
+          <broadcast></broadcast>
```
Comment (Klaas-) [inline provisioning_templates/provision/autoyast_default.erb]: I don't think you'll need this line

File: provisioning_templates/provision/autoyast_default.erb [hunk 3]
```diff
@@ -22,21 +22,57 @@ oses:
...
         <search><%= @host.domain %></search>
       </searchlist>
-<% end -%>
+    <% end -%>
     </dns>
     <interfaces config:type="list">
       <interface>
-        <bootproto>dhcp</bootproto>
         <device>eth0</device>
+        <ethtool_options>autoneg on</ethtool_options>
```
Comment (mmoll) [inline provisioning_templates/provision/autoyast_default.erb]: is this line really needed?

File: provisioning_templates/provision/autoyast_default.erb [hunk 4]
```diff
@@ -22,21 +22,57 @@ oses:
...
         <usercontrol>no</usercontrol>
       </interface>
     </interfaces>
+    <% if @host.subnet.respond_to?(:dhcp_boot_mode?) -%>
+    <routing>
+    <ip_forward config:type="boolean">false</ip_forward>
+    <routes config:type="list">
+      <route>
+        <destination>default</destination>
+        <device>eth0</device>
```
Comment (mmoll) [inline provisioning_templates/provision/autoyast_default.erb]: I guess (but we do have the same problem for DHCP also, anyway...) that's not always correct with persistent device naming in newer versions?

File: provisioning_templates/provision/autoyast_default.erb [hunk 5]
```diff
@@ -22,21 +22,57 @@ oses:
...
+        <gateway><%= @host.subnet.gateway %></gateway>
+        <netmask>-</netmask>
```
Comment (mmoll) [inline provisioning_templates/provision/autoyast_default.erb]: what does the dash mean?

File: provisioning_templates/provision/autoyast_default.erb [hunk 6]
```diff
@@ -22,21 +22,57 @@ oses:
...
-<% end -%>
+    <% end -%>
     </dns>
     <interfaces config:type="list">
       <interface>
-        <bootproto>dhcp</bootproto>
         <device>eth0</device>
+        <ethtool_options>autoneg on</ethtool_options>
+        <% if @host.subnet.respond_to?(:dhcp_boot_mode?) -%>
+          <bootproto>static</bootproto>
```
Comment (ekohl) [inline provisioning_templates/provision/autoyast_default.erb]: The XML indenting will look weird here.

File: provisioning_templates/provision/autoyast_default.erb [hunk 7]
```diff
@@ -22,21 +22,57 @@ oses:
...
       </searchlist>
-<% end -%>
+    <% end -%>
     </dns>
     <interfaces config:type="list">
       <interface>
-        <bootproto>dhcp</bootproto>
         <device>eth0</device>
+        <ethtool_options>autoneg on</ethtool_options>
+        <% if @host.subnet.respond_to?(:dhcp_boot_mode?) -%>
```
Comment (ekohl) [inline provisioning_templates/provision/autoyast_default.erb]: If it responds to boot mode but boot mode is set to DHCP then you will still end up with static networking.

**Review by ekohl — CHANGES_REQUESTED** [review]

File: provisioning_templates/provision/autoyast_default.erb (hunk 7 above)
Comment (hboetes) [reply provisioning_templates/provision/autoyast_default.erb]: I c&p this line from the kickstart template, what do you propose I should use?

File: provisioning_templates/provision/autoyast_default.erb (hunk 5 above)
Comment (hboetes) [reply provisioning_templates/provision/autoyast_default.erb]: I don't know, but I find it all over the docs. https://www.suse.com/documentation/sles-12/book_autoyast/data/createprofile_network.html

File: provisioning_templates/provision/autoyast_default.erb (hunk 3 above)
Comment (hboetes) [reply provisioning_templates/provision/autoyast_default.erb]: That depends on the default configuration in autoyast. autoneg is what most people will want to have.

Comment (mmoll) [reply provisioning_templates/provision/autoyast_default.erb]: Isn't autoneg the default? So this line can get ommited?

File: provisioning_templates/provision/autoyast_default.erb (hunk 5 above)
Comment (mmoll) [reply provisioning_templates/provision/autoyast_default.erb]: ah, I see 👍

File: provisioning_templates/provision/autoyast_default.erb (hunk 4 above)
Comment (hboetes) [reply provisioning_templates/provision/autoyast_default.erb]: Fixed in my latest commit which supports multiple interfaces

File: provisioning_templates/provision/autoyast_default.erb (hunk 6 above)
Comment (hboetes) [reply provisioning_templates/provision/autoyast_default.erb]: I noticed as well. What can I do about it?

**Comment by mmoll** [comment]
Regarding `dhcp_boot_mode`, the `respond_to?` is only for identifying if the Foreman in question does support the distinction. As the templates are now versioned, this can/should be ommited.

Regarding the distinction between DHCP and static mode, here is the simple code used in the preseed networking snippet:
```
<% @host.managed_interfaces.each do |interface| -%>
  <% next if !interface.managed? || interface.subnet.nil? || interface.primary -%>
  <% subnet = interface.subnet -%>
  <% dhcp = subnet.nil? ? false : subnet.dhcp_boot_mode? -%>
  <% if dhcp -%>
    [do dhcp things here]
  <% else -%>
    [do static things here]
  <% end -%>
<% end -%>
```

File: provisioning_templates/provision/autoyast_default.erb (hunk 6 above)
Comment (mmoll) [reply provisioning_templates/provision/autoyast_default.erb]: indenting it correctly? ;)

File: provisioning_templates/provision/autoyast_default.erb (hunk 3 above)
Comment (hboetes) [reply provisioning_templates/provision/autoyast_default.erb]: ok, consider it gone.

**Comment by mmoll** [comment]
Again, `<% if @host.subnet.respond_to?(:dhcp_boot_mode?) -%>` does _not_ check if that subnet is in DHCP or static mode.

You copypasted some code from the kickstart snippet, but in your the current code there's no usage of the variables `virtual`, `vlan`, `alias-type` and even `dhcp` of the respective subnet. Please double-check your code and also make sure, DHCP installs work as intended.

**Review by mmoll — CHANGES_REQUESTED** [review]
this needs to get into a working state first

**Comment by hboetes** [comment]
If you guys want me to join all these commits, I'll gladly oblige.

Also, I'd like to note that travis-ci only tests kickstart files.

**Comment by mmoll** [comment]
@hboetes thanks for following up on this, looking much better now.

I squashed all your commit together into one and then added one commit with some refactorings.

This is still untested and I don't know when I will have time to test this with real machines, in the meanwhile I'd ask @ares to have a look. 👀

**Comment by Klaas-** [comment]
I think there is another issue here we should address: the device identifiers are not enforced by foreman at the moment, if they aren't set the network interfaces won't be configured  properly (at least if I didn't mess up my test case :) ). I'm not sure how one would work around that - could we just set a device name based on a loop-counter or can we do it like the rhel networking snippet does - set no device identifiers and use the mac to identify the device?

**Comment by mmoll** [comment]
I did think about that, but at the end this is already an improvement over the current situation, where eth0 is also hardcoded. I hope to test this in the next days and merge it, further improvements shoudl go into further PRs.

**Comment by mmoll** [comment]
merged, thanks @hboetes!

**Pull request merged**